Mathlib/LinearAlgebra/Multilinear/Basic.lean

```diff
@@ -760,8 +763,165 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
 
 end
 
+
+section Derivative
+
+/-! The goal of this part is, for `f` a multilinear map in finitely many variables indexed by `ι`,
+to develop `f(x + y)` as a finite formal multilinear series in `y`. The term of degree `n` of
+this formal multilinear series will be a sum over finsets `s` of `ι` of cardinality `n` of a
+multilinear map indexed by `s`, evaluated at `fun (_ : s) => y`; we use a linear order on
+`ι` to identify all such `s` to `Fin n`. We also give a more direct definition of the term
+of degree `1` in `f.linearDeriv`, which doesn't require a linear order on `ι`, and prove the
+equivalence of the two definitions.-/
+
+lemma domDomRestrict_aux [DecidableEq ι] (s : Set ι) [(i : ι) → Decidable (i ∈ s)]
+    [DecidableEq s] (x : (i : ι) → M₁ i) (z : (i : s) → M₁ i) (i : s)
+    (c : M₁ i) : (fun j ↦ if h : j ∈ s then Function.update z i c ⟨j, h⟩ else x j) =
+    Function.update (fun j => if h : j ∈ s then z ⟨j, h⟩ else x j) i c := by
+  ext j
+  by_cases h : j = i.1
+  · rw [h, Function.update_same]
+    simp only [Subtype.coe_prop, Function.update_same, dite_eq_ite, ite_true]
+  · rw [Function.update_noteq h]
+    by_cases h' : j ∈ s
+    · simp only [h', ne_eq, dite_true]
+      have h'' : ¬ ⟨j, h'⟩ = i :=
+        fun he => by apply_fun (fun x => x.1) at he; exact h he
+      rw [Function.update_noteq h'']
+    · simp only [h', ne_eq, dite_false]
+
+/-- Given a multilinear map `f` on `(i : ι) → M i`, an element `x` of `(i : ι) → M i` and s
+set `s` of `ι`, construct a multilinear map on `s → ((i : ι) → Mi)` whose value at `z`
+is `f` evaluated at the vector with `i`th coordinate `(z i) i` if `i ∈ s` and `x i` otherwise.-/
+def domDomRestrict [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : ι) → M₁ i)
+    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
+    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
+  toFun z := f (fun i ↦ if h : i ∈ s then z ⟨i, h⟩ else x i)
```

Comment:
How much messier do things become if you generalize to
```suggestion
def domDomRestrict [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : sᶜ) → M₁ i)
    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
  toFun z := f (fun i ↦ if h : i ∈ s then z ⟨i, h⟩ else x ⟨i, h⟩)
```

Reply:
Not sure. I'm nervous because of things like [this](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Multilinear/Basic.html#MultilinearMap.toLinearMap), where you take `m` defined on all of `ι` even though you don't actually have to. (Mathematically it doesn't matter what we choose here.)

Reply:
Doesn't seem to change much. I can make the change, I'm pretty indifferent to it, just not sure what is more compatible with standard mathlib practice.

Reply:
Made the change.